CAM: Improve warning text for improper unit scheme

jbaehr opened the pull request:
This PR is a response to #19441, following up #11851. While there is no bug in a technical sense, I think we can still improve the text of improper unit warning, to not guide our users into the wrong direction:
The previous text suggested to change the scheme in the preferences, however, the preferences only take effect for new documents. Now, that we have the "ProjectUnitSystem" in place, the setting for the *current* document needs to be changed.

I also removed some explicit newlines, which seem to be counterproductive as the dialog already does automatic line breaks.

Previously:
![image](https://github.com/user-attachments/assets/1f0a21cf-8587-4e0a-a92a-766a5c2f2af1)

With this PR:
![image](https://github.com/user-attachments/assets/d48117d5-e3b4-4a1c-ad1d-75e5fb2ef591)


*edit*: update image to latest changes

JohnOCFII commented:
Two comments:

1) In the third line, you could make the word "does" lower case as it doesn't start a new sentence.
2) Any chance you could modify the text in the "Don't Show this Anymore" box?  I think the text, "Don't show this message again" (note the case change - not capitalizing each word). If that takes too much space, you could also shorten to, "Don't show again"

jbaehr commented:
@JohnOCFII thanks for the review.
> 1. In the third line, you could make the word "does" lower case as it doesn't start a new sentence.

Agreed and addressed.

> 2. Any chance you could modify the text in the "Don't Show this Anymore" box?  I think the text, "Don't show this message again" (note the case change - not capitalizing each word). If that takes too much space, you could also shorten to, "Don't show again"

Yes, this text can easily be changed; it's just a few lines below my changes.
I don't have any strong feelings here, but I do value consistency. Do you have other examples in FreeCAD where such buttons are used? How to they handle the situation?
@luzpaz, what's your take on this?

JohnOCFII commented:
> I don't have any strong feelings here, but I do value consistency. Do you have other examples in FreeCAD where such buttons are used? How to they handle the situation?

From the Start page:
<img width="1642" alt="Screenshot 2025-02-17 at 4 20 51 PM" src="https://github.com/user-attachments/assets/3701947e-0947-4e7d-86b3-3a96e53be5ec" />


It is wordier than I'd like, but flows similarly to what I suggested above.

adrianinsaval commented:
pardon my ignorance but why can't CAM just convert values when required? It is very odd that a specific unit schema is required

jbaehr commented:
I changed the button text as per @JohnOCFII's suggestion. It now follows the "Don't show this XYZ again" pattern we have on the Start page. On a mac with Qt5 it now looks like this:
![image](https://github.com/user-attachments/assets/c84b13cc-aba1-4e0e-b7ed-cc9c242e1fb8)

hyarion commented:
> pardon my ignorance but why can't CAM just convert values when required? It is very odd that a specific unit schema is required

I've been wondering about this too, feels like a bug, but I might miss something?

Anyway, I think this fix is good in the meantime.

JohnOCFII commented:
> I've been wondering about this too, feels like a bug, but I might miss something?

@sliptonic might be the best person to give us the reason it operates this way in FreeCAD today.

hyarion commented:
> > I've been wondering about this too, feels like a bug, but I might miss something?
> 
> 
> 
> Same here. I thought this is something for another day, though.

I agree.

sliptonic commented:
> > I've been wondering about this too, feels like a bug, but I might miss something?
> 
> @sliptonic might be the best person to give us the reason it operates this way in FreeCAD today.

Gcode is unitless for velocity.  The G20/G21 doesn't touch it.  FreeCAD uses explicit units.  The default FreeCAD condition (unit/sec) is opposite to the normal situation in the CAM world (unit/min).

The resulting situation will almost always lead do a dangerous situation.

Assume an experienced CAM programmer new to FreeCAD sets up a job.  He enters the Feed rate as '200' assuming it is in units/min. (Extremely easy mistake to make) .  Then he postprocesses the job.  All the postprocessors convert to unit/min because that's the only thing gcode takes.  So the output will be converted to unit/min and will be 200*60 =12000 unit/sec.   This is a crash situation that can damage the machine or injure the operation.

The decision was to warn the user that they are not using a unit schema consistent with gcode.

jbaehr commented:
> The decision was to warn the user that they are not using a unit schema consistent with gcode.

I see... so it's not that we'd get "incorrect" gcode. We do have [proper conversion](https://github.com/FreeCAD/FreeCAD/blob/dda05369c4705759926ebfc66412e66675380b84/src/Mod/CAM/Path/Post/UtilsParse.py#L517), depending on G20/G21 into "inch/min" or "mm/min", and since #18685 also "°/min". It's rather that the values *displayed in FreeCAD* (e.g. in the properties) match what ends up written into gcode.

Then we should rephrase the warning, to avoid the question marks we all had in this discussion above before you jumped in.

What do you think about changing this
> CNC machines require feed rate to be expressed in unit/minute. To ensure correct G-code, select a minute-based schema in the document properties.

into 
> In G-code, feed rates are expressed in unit/minute. To ensure consistency with the GUI, select a minute-based schema in the document properties.

or
> In G-code, feed rates are expressed in unit/minute. To use the same in FreeCAD's GUI, select a minute-based schema in the document properties.

sliptonic commented:
> > The decision was to warn the user that they are not using a unit schema consistent with gcode.
> 
> I see... so it's not that we'd get "incorrect" gcode. We do have [proper conversion](https://github.com/FreeCAD/FreeCAD/blob/dda05369c4705759926ebfc66412e66675380b84/src/Mod/CAM/Path/Post/UtilsParse.py#L517), depending on G20/G21 into "inch/min" or "mm/min", and since #18685 also "°/min". It's rather that the values _displayed in FreeCAD_ (e.g. in the properties) match what ends up written into gcode.

Not quite.  The user will enter '200' and get 12000.   The schema controls how the values are displayed and also how they are interpreted on input.  If the user is working in US standard and puts '1' in a length field, it will be understood as 1 inch,  shown as '1' and stored as 25.4 (mm).  

The feed rate field is of type velocity.  If the schema is expecting unit/sec  and the user buts in '200', it will be interpreted as 200 units/second.  Since we need to output units per minute, the conversion will result in 12000 and the machine will crash.

I'm ok with changing the warning to something that you think is more easily understood.  I have no strong feeling about the wording.  The better solution would be to present the user with a dialog that explains the problem and also gives them a selection control to change the schema right there.   Only populate the choices with schemas that support  unit/min for velocity.

hyarion commented:
@sliptonic wouldnt it be preferred to have input fields in CAM that always work in units/minute. So if you enter a unitless quantity it would give you mm/minute even if you use mm/s as unit?

jbaehr commented:
> ... wouldnt it be preferred to have input fields in CAM that always work in units/minute. So if you enter a unitless quantity it would give you mm/minute even if you use mm/s as unit?

I don't think this is a good idea. I mean, if you select a minute-based unit scheme, then you have exactly this. Everywhere. Having it scheme-independent only in CAM, would mean that you end up with different input methods when switching e.g. between CAM and Sketcher. Very confusing.

jbaehr commented:
> The schema controls how the values are DISPLAYED and also how they are INTERPRETED on input.

Yes, I did those tests before I came up with the other warning text proposals. I did not update the upper part of my post, though.

> The better solution would be to present the user with a dialog that explains the problem and also gives them a selection control to change the schema right there.

Agreed. Yet another item for the ever growing TODO list ;-)
For now I'd like to go with 
> In G-code, feed rates are expressed in unit/minute. To use the same in FreeCAD's GUI, select a minute-based schema in the document properties.

I think this covers input and output elements alike.

sliptonic commented:
> > ... wouldnt it be preferred to have input fields in CAM that always work in units/minute. So if you enter a unitless quantity it would give you mm/minute even if you use mm/s as unit?
> 
> I don't think this is a good idea. I mean, if you select a minute-based unit scheme, then you have exactly this. Everywhere. Having it scheme-independent only in CAM, would mean that you end up with different input methods when switching e.g. between CAM and Sketcher. Very confusing.

I don't think it's even feasible.  Sure you could make your own input controls in the task panels but the property panes would still use the freecad core input widgets which are unit-aware.  So the user would get one behavior if they were in a task panel and  a different behavior in the property pane.  Confusing.

sliptonic commented:
I don't know if the FEM folks (or anyone else) uses velocity.   Not exactly common in CAD.  
We could make the argument that the default schema that ships with FreeCAD should be minute-based.

hyarion commented:
> I don't think it's even feasible. Sure you could make your own input controls in the task panels but the property panes would still use the freecad core input widgets which are unit-aware. So the user would get one behavior if they were in a task panel and a different behavior in the property pane. Confusing.

I think it depends on presentation. People are already confused and tries to use the "wrong unit system" even though they want to have different systems normally (and another one in CAM).

Another option would be to allow document objects to override unit systems as described here: https://forum.freecad.org/viewtopic.php?p=784257#p784257

jbaehr commented:
> Another option would be to allow document objects to override unit systems ...

That'll be very interesting, indeed. But on a property-level: Even within a single CAM document object, I'd need
- µm for the chip load, opposed to mm for everything else. (I don't need more than two sub-mm digits, but then 0.025 mm gets rounded to 0.03 for display)
- mm/min for the feed rate. Obviously.
- m/s for the cutting speed (a custom property I usually add to my tool controllers, used in expressions to automatically calculate the spindle speed and indirectly the feed rate)
- 1/min for the spindle speed (again a custom prop -- the build-in spindle speed is unfortunately a unit-less float, spoiling expressions)

sliptonic commented:
> Another option would be to allow document objects to override unit systems as described here: https://forum.freecad.org/viewtopic.php?p=784257#p784257

I suspect that is pretty far down the road.  A lot of work.  A lot of edge-cases, A lot of testing needed, and probably pretty far down the priority list.

The near term priority should be 
1) a clear and helpful message.
2) a link to a wiki page that explains it more.   (Based on this PR discussion, it's always going to be confusing)
3) A reworked panel that lets the user fix the user schema right there.

With regard to the current panel, I know I said I didn't have strong feelings but one thing jumps out at me:
The message says, "the current document does not use minutes for velocity".   This is contrasted where it explains the requirements of gcode and explains it correctly as unit/time.

I think those two parts of the message should use the same unit expression and both be worded in the affirmative.  Something like:

"The current document unit schema expresses velocity as mm/SECOND.  

CNC machines expect feed rate velocity to be expressed as unit/MINUTE. 

... "

jbaehr commented:
> I don't know if the FEM folks (or anyone else) uses velocity. Not exactly common in CAD. We could make the argument that the default schema that ships with FreeCAD should be minute-based. That doesn't solve the problem but minimizes how many people even see the warning messages

We could advertise the unit scheme "Metric, Small Parts & CNC" as default in the first-run wizard. I assume that's best our largest user base.
But beyond that I'd be carful. I'm not a FEM guy, but in physics simulations you use m/s and rad/s (at least MBDyn does, no idea about the amateur rocketeers). Perhaps in the transportation WB prefers km/h... I'd say beyond CNC velocities in per-minute units are quite unusual. End even in this domain I've seen mm/s for feeds in German documentation (see for example [this spindle manual](https://www.stepcraft-systems.com/download/20211111_HF500_DE_EN.pdf), page 21)

hyarion commented:
I agree that the discussion about changing units for objects/input fields are down the line. Improving the message is enough for this pr

jbaehr commented:
So back to topic. Here is the next proposal:
![image](https://github.com/user-attachments/assets/95c42115-233d-4cf9-892e-fc299481efc4)

I also thought about writing 'minutes' and 'seconds' capitalized, but [the NIST tells otherwise](https://www.nist.gov/pml/owm/writing-si-metric-system-units#capitalization).

LarryWoestman commented:
I think you would want "expresses velocity in values per second" for idiomatic english.
I have doubts about US Customary being a useful choice if you are going to be designing parts that will result in Gcode.  I suggest leaving it out of the list.

sliptonic commented:
The capitalization isn't to be NIST compliant, it's to draw the user's attention to the distinction.  That's a long error message and it's tempting to just click ok, or better yet, 'don't show me again'   rather than read.

The wording also makes it sound like this is a nice GUI thing that will make your life better.  I think the warning should be stronger.  Using an improper unit schema can result in dangerous gcode and machine crashes.  You have been warned.  Maybe with a skull icon and printed in a really [scary font](https://www.dafont.com/curse-of-the-zombie.font).

I'm kidding about that last bit, but the message should be a stern warning

hyarion commented:
Maybe start it with "Incompatible unit schema!" (or something similar) and set the rest with setInformativeText to get different sized text to highlight the importance of the first part. If all text is the same size, you only see a blob.

![image](https://github.com/user-attachments/assets/5bb42f6b-5964-4547-bbf3-c66b7aa71e60)

jbaehr commented:
> The capitalization isn't to be NIST compliant, it's to draw the user's attention to the distinction. 

Sorry, I misinterpreted your comment above. I thought, the ALL CAPS was to draw *my* attention to the "use the same unit expression" part, not to have it written like that in the message box. (I should have looked up what "worded in the affirmative" means, first.).
My comment about the NIST was regarding "Minutes" vs "minutes".

![image](https://github.com/user-attachments/assets/bf5f34cf-dfc2-4c22-a73e-66b71c00a08e)

Here comes the next proposal. Currently using HTML, but now that we require at least Qt-5.15 using Markdown also becomes an option. Not sure what our translators think about that, though.

Still need to clean the surrounding code a bit -- *eight* levels of indentation is a bit too much.

jbaehr commented:
I had a short chat with a UX guy at my day job (neither CAD nor CAM related) about this.
Actually, a message like this has to cover these three points:
- What is the problem
- How can I fix it
- What happens if I don't

While those are addressed in my last proposal, his assessment was as well that  the message itself is way too long/complex. Other phrasing is welcome, especially from native speakers.

Another point he brought up is that if this issue is so severe that it can crash your machine, we should probably *not* offer a "don't show this again" option. Or at least restrict this to document in question, not as a global option.

sliptonic commented:
I like the clarity of the three-points.  Here's my attempt at simplifying:

"This document uses a unit schema that expresses velocity _per second_. This  can cause a dangerous situation or crash your CNC machine

Select a unit schema that expresses feed rates _per minute_ rather than _per second_.  For example:
- Metric small parts & CNC (mm, mm/min)
- Imperieal decimal (in, lb)

[more information](URL).
"

chennes commented:
@jbaehr I think the ball is in your court here, on making or adapting @sliptonic's suggested wording.

jbaehr commented:
@chennes yes it is. I was quite busy with non-FreeCAD stuff lately, but I'll pick it up during this week.

jbaehr commented:
I hope I have incorporated the feedback from all comments now. This is what it looks like on a mac (where the title [is ignored](https://doc.qt.io/qt-6/qmessagebox.html#setWindowTitle)):
![image](https://github.com/user-attachments/assets/d48117d5-e3b4-4a1c-ad1d-75e5fb2ef591)

I removed all the technical details and added a link to the wiki. HTML is used for the emphasis.

I also refactored the code in a separate commit.
